src/basilisp/lang/compiler/exception.py

```diff
@@ -40,40 +46,55 @@ class _loc:
     def __bool__(self):
         return (
             self.line is not None
-            and self.col is not None
-            and self.end_line is not None
-            and self.end_col is not None
+            or self.col is not None
+            or self.end_line is not None
+            or self.end_col is not None
         )
 
 
 @attr.define(str=False)
 class CompilerException(IExceptionInfo):
     msg: str
     phase: CompilerPhase
+    filename: Optional[str] = None
     form: Union[LispForm, None, ISeq] = None
     lisp_ast: Optional[Node] = None
     py_ast: Optional[ast.AST] = None
 
     @property
     def data(self) -> IPersistentMap:
         d: Dict[kw.Keyword, Any] = {_PHASE: self.phase.value}
+        if self.filename is not None:
+            d[_FILE] = self.filename
         loc = None
         if self.form is not None:
             d[_FORM] = self.form
             loc = (
                 _loc(
                     self.form.meta.val_at(READER_LINE_KW),
                     self.form.meta.val_at(READER_COL_KW),
+                    self.form.meta.val_at(READER_END_LINE_KW),
+                    self.form.meta.val_at(READER_END_COL_KW),
                 )
                 if isinstance(self.form, IMeta) and self.form.meta
                 else None
             )
         if self.lisp_ast is not None:  # pragma: no cover
             d[_LISP_AST] = self.lisp_ast
-            loc = loc or _loc(self.lisp_ast.env.line, self.lisp_ast.env.col)
+            loc = loc or _loc(
+                self.lisp_ast.env.line,
+                self.lisp_ast.env.col,
+                self.lisp_ast.env.end_line,
+                self.lisp_ast.env.end_col,
+            )
         if self.py_ast is not None:  # pragma: no cover
             d[_PY_AST] = self.py_ast
-            loc = loc or _loc(self.py_ast.lineno, self.py_ast.col_offset)
+            loc = loc or _loc(
+                self.py_ast.lineno,
+                self.py_ast.col_offset,
+                self.py_ast.end_lineno,
+                self.py_ast.end_col_offset,
```

Comment:
These were added in Python 3.8 and are therefore now guaranteed in all supported versions of Python.